kairon/shared/data/processor.py

```diff
@@ -1132,6 +1182,86 @@ def __prepare_training_story_events(self, events, timestamp):
+        hubspot_forms_action = dict(HubspotFormsAction.objects(bot=bot, status=True).values_list('name', 'id'))
+        two_stage_fallback_action = dict(KaironTwoStageFallbackAction.objects(bot=bot, status=True).values_list('name', 'id'))
+        prompt_action = dict(PromptAction.objects(bot=bot, status=True).values_list('name', 'id'))
+        web_search_action = dict(WebSearchAction.objects(bot=bot, status=True).values_list('name', 'id'))
+        all_actions = [http_action, google_search_action, email_action, slot_set_action, jira_action, form_action,
+                       zendesk_action, pipedrive_leads_action, hubspot_forms_action, two_stage_fallback_action,
+                       prompt_action, web_search_action]
+        combined_actions = {key: value for action in all_actions for key, value in action.items()}
+        actions['actions'] = combined_actions
+        return component_dict, actions
+
+    def __updated_events(self, bot, events):
+        component_dict, actions = self.__retrieve_existing_components(bot)
+        for event in events:
+            step = event['step']
+            connections = event.get('connections', [])
+            step_type = step.get('type')
+            step_name_lower = step.get('name').lower()
+
+            if step_type in component_dict:
+                step['component_id'] = str(component_dict[step_type].get(step_name_lower))
+            elif step_name_lower in actions['actions']:
+                step['component_id'] = str(actions['actions'].get(step_name_lower))
+            if connections:
+                for connection in connections:
+                    connection_type = connection['type']
+                    connection_name_lower = connection['name'].lower()
+                    if connection_type in component_dict:
+                        connection['component_id'] = str(component_dict[connection_type].get(connection_name_lower))
+                    elif connection_name_lower in actions['actions']:
+                        connection['component_id'] = str(actions['actions'].get(connection_name_lower))
+        return events
+
+    def __fetch_multiflow_story_block_names(self, bot: Text):
+        multiflow_stories = list(
+            MultiflowStories.objects(bot=bot, status=True).values_list('block_name')
+        )
+        return multiflow_stories
+
+    def __extract_multiflow_story_step(self, multiflow_stories, bot: Text, user: Text):
+        existing_multiflow_stories = self.__fetch_multiflow_story_block_names(bot)
+        existing_stories = self.__fetch_story_block_names(bot)
+        existing_rules = self.fetch_rule_block_names(bot)
+        collection = set(existing_multiflow_stories + existing_stories + existing_rules)
+        for story in multiflow_stories:
+            if story['block_name'].strip().lower() not in collection:
+                story['events'] = self.__updated_events(bot, story['events'])
```

Comment:
we should fetch the existing components and apply for each story instead of retrieving components for each story